spec/compromise-syntax.md

```diff
@@ -0,0 +1,238 @@
+and after the function name there must be only white-space-separated options which
+start with identifier-start characters.
+
+# Syntactic white space
+
+We use white space inside placeholders and in “code mode” (outside patterns) as token separators.
+White space is a sequence of one or more of the characters TAB, LF, CR, SP, and maybe some more.
+- TODO: For the definition of white space we should consult with the Unicode Source Code Working Group.
+- TODO: Decide whether to use Unicode Pattern_White_Space or otherwise allow RLM and LRM characters.
+
+White space can also be useful for line breaking long messages, indentation, and alignment.
+However, we should not allow white space everywhere possible,
+because that just leads to confusing variations in style,
+and the creation of formatting tools to enforce certain styles.
+For example, there is no reason to allow white space between a name or function prefix and its identifier,
+around the `=` of an option, after the `{` of a placeholder, or before the `}` of a placeholder.
+
+# Pattern selection
+
+Messages need the ability to choose among variants of a pattern based on certain argument values.
+Common examples include selecting the right plural form, and variants for different person genders.
+
+There should be a single level of selection (not nested like in ICU MessageFormat).
+It needs to support multiple selectors.
+
+In this syntax, a list of N selectors is followed by a list of pairs where
+the first element of each pair is a list of N value literals and
+the second element of each pair is a pattern.
+A `_` is a wildcard value that always matches.
+The last variant must have a list of all wildcard values.
+```
+[{$count :plural offset=1 grouping=always} {$gender}]
+[1 female] {{$name} added you to her circles.}
+[1 male] {{$name} added you to his circles.}
+[1 _] {{$name} added you to their circles.}
+[_ _] {{$name} added you and {#count} others to their circles.}
+```
+
+Lists are enclosed in square brackets, reminiscent of Python lists.
+The opening `[` also distinguishes the selection syntax from a simple pattern.
+
+TODO: Decide whether to enclose each value literal in
+the same pair of delimiters as literals in placeholder (for consistency),
+or whether to make that optional.
+(The `[]` value list syntax already indicates that value literals are enclosed.)
+Some literals may require it if they contain spaces.
+The `_` should probably never be enclosed in literal delimiters.
+
+Selector syntax follows placeholder syntax,
+except that a function must be specified.
+For the purpose of selection, there are three types of functions:
+1. Select-and-format functions combine the two functionalities,
+   and the selection is informed by the formatting.
+   For example, selectors for plural variants
+   (different selectors for cardinal-number vs. ordinal-number variants)
+   have to take into account how the number is formatted.
+2. Format-only functions can be used as selectors via
```

Comment:
but only if the output doesn't contain spaces?

Reply:
We can decide to support spaces by allowing or requiring delimiters around the variant values.